File: app/backend/app/schemas/lead.py
from __future__ import annotations

import json
import re
from datetime import date, datetime
from typing import Any, Optional

from pydantic import BaseModel, field_validator, model_validator

from app.models.lead import LeadSourceType, LeadStatus, ServiceType
from app.schemas.followup import FollowupOut

_SOURCE_LABELS: dict[str, str] = {
    "thumbtack_api":        "Thumbtack API",
    "thumbtack_screenshot": "Thumbtack Screenshot",
    "yelp_screenshot":      "Yelp Screenshot",
    "google_screenshot":    "Google Screenshot",
    "website_form":         "Website Form",
    "manual":               "Manual Entry",
}

_TIME_SLOT_RE = re.compile(r"^(?:[01]\d|2[0-3]):[0-5]\d$")


def _normalize_time_slot(value: Any) -> Optional[str]:
    if value is None:
        return None
    if not isinstance(value, str):
        raise ValueError("appointment_time_slot must be a string in HH:MM format")
    value = value.strip()
    if not value:
        return None
    if not _TIME_SLOT_RE.fullmatch(value):
        raise ValueError("appointment_time_slot must be in HH:MM format")
    return value


def _normalize_duration_minutes(value: Any) -> Optional[int]:
    if value is None:
        return None
    if isinstance(value, str):
        value = value.strip()
        if not value:
            return None
        if not value.isdigit():
            raise ValueError("estimated_job_duration_minutes must be a whole number of minutes")
        value = int(value)
    elif isinstance(value, bool) or not isinstance(value, int):
        raise ValueError("estimated_job_duration_minutes must be a whole number of minutes")

    if value <= 0:
        raise ValueError("estimated_job_duration_minutes must be greater than 0")
    if value > 24 * 60:
        raise ValueError("estimated_job_duration_minutes must be 1440 minutes or less")
    return value


class QuoteModifierIn(BaseModel):
    amount: float
    note: str


class QuoteModifierOut(QuoteModifierIn):
    pass


class LeadCreate(BaseModel):
    city_id: Optional[str] = None
    source_type: LeadSourceType
    source_reference_id: Optional[str] = None
    raw_payload: Optional[str] = None
    customer_name: str
    customer_phone: Optional[str] = None
    service_type: ServiceType = ServiceType.unknown
    job_location: Optional[str] = None
    job_origin: Optional[str] = None
    job_destination: Optional[str] = None
    job_date_requested: Optional[date] = None
    appointment_time_slot: Optional[str] = None
    estimated_job_duration_minutes: Optional[int] = None
    scope_notes: Optional[str] = None
    # Intake notes — initial context recorded at lead creation
    notes: Optional[str] = None
    assigned_to: Optional[str] = None
    urgency_flag: bool = False
    # Slice 8
    move_distance_miles: Optional[float] = None
    load_stairs: Optional[int] = None
    unload_stairs: Optional[int] = None
    move_size_label: Optional[str] = None
    move_type: Optional[str] = None
    move_date_options: Optional[list[str]] = None
    job_date_end: Optional[str] = None
    job_date_end: Optional[str] = None
    accept_and_pay: bool = False

    _validate_appointment_time_slot = field_validator("appointment_time_slot", mode="before")(
        _normalize_time_slot
    )
    _validate_duration_minutes = field_validator("estimated_job_duration_minutes", mode="before")(
        _normalize_duration_minutes
    )


class LeadUpdate(BaseModel):
    """Partial update for mutable lead fields. Only supplied fields are written.
    For operational notes logged during live handling, use POST /leads/{id}/notes."""
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None
    service_type: Optional[ServiceType] = None
    job_location: Optional[str] = None
    job_origin: Optional[str] = None
    job_destination: Optional[str] = None
    job_date_requested: Optional[date] = None
    appointment_time_slot: Optional[str] = None
    estimated_job_duration_minutes: Optional[int] = None
    scope_notes: Optional[str] = None
    urgency_flag: Optional[bool] = None
    assigned_to: Optional[str] = None
    # Intake notes only — corrections to initial context, not live handling updates
    notes: Optional[str] = None
    # Slice 8 / 9
    move_distance_miles: Optional[float] = None
    load_stairs: Optional[int] = None
    unload_stairs: Optional[int] = None
    move_size_label: Optional[str] = None
    move_type: Optional[str] = None
    move_date_options: Optional[list[str]] = None
    quote_context: Optional[str] = None
    job_date_end: Optional[str] = None
    # Confirmed physical address — setting this triggers auto-booking
    job_address: Optional[str] = None

    _validate_appointment_time_slot = field_validator("appointment_time_slot", mode="before")(
        _normalize_time_slot
    )
    _validate_duration_minutes = field_validator("estimated_job_duration_minutes", mode="before")(
        _normalize_duration_minutes
    )


class NoteCreate(BaseModel):
    """Operational note appended to the lead event log during live handling."""
    body: str
    actor: Optional[str] = None


class LeadStatusUpdate(BaseModel):
    status: LeadStatus
    actor: Optional[str] = None
    note: Optional[str] = None
    quoted_price_total: Optional[float] = None
    quote_modifiers: Optional[list[QuoteModifierIn]] = None
    estimated_job_duration_minutes: Optional[int] = None

    _validate_duration_minutes = field_validator("estimated_job_duration_minutes", mode="before")(
        _normalize_duration_minutes
    )


class LeadEventOut(BaseModel):
    id: str
    lead_id: str
    event_type: str
    from_status: Optional[str]
    to_status: Optional[str]
    actor: Optional[str]
    note: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class ScreenshotOut(BaseModel):
    id: str
    lead_id: str
    original_filename: str
    stored_path: str  # relative path; prefix with /uploads/ to build URL
    file_size: int
    ocr_status: Optional[str] = None
    # "intake", "correspondence", "before_job", "after_job"
    screenshot_type: str = "intake"
    created_at: datetime

    model_config = {"from_attributes": True}


class LeadOut(BaseModel):
    id: str
    city_id: str
    city_name: Optional[str] = None
    city_slug: Optional[str] = None
    source_type: LeadSourceType
    source_reference_id: Optional[str]
    raw_payload: Optional[str]
    status: LeadStatus
    urgency_flag: bool
    # None for ingest stubs not yet confirmed by OCR or the facilitator
    customer_name: Optional[str]
    customer_phone: Optional[str]
    service_type: ServiceType
    job_location: Optional[str]
    job_origin: Optional[str] = None
    job_destination: Optional[str] = None
    job_date_requested: Optional[date]
    appointment_time_slot: Optional[str] = None
    estimated_job_duration_minutes: Optional[int] = None
    scope_notes: Optional[str] = None
    # JSON dict: {field_name: "ocr" | "edited"} — parsed by the router; absence = manual
    field_sources: Optional[dict[str, Any]] = None
    notes: Optional[str]
    ingested_by: Optional[str] = None
    assigned_to: Optional[str]
    created_at: datetime
    acknowledged_at: Optional[datetime]
    updated_at: datetime
    # Slice 8 / 9
    move_distance_miles: Optional[float] = None
    load_stairs: Optional[int] = None
    unload_stairs: Optional[int] = None
    move_size_label: Optional[str] = None
    move_type: Optional[str] = None
    move_date_options: Optional[list[str]] = None
    accept_and_pay: bool = False
    quote_context: Optional[str] = None
    quoted_price_total: Optional[float] = None
    quote_modifiers: Optional[list[QuoteModifierOut]] = None
    job_address: Optional[str] = None
    job_date_end: Optional[str] = None
    dispatched_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    # Computed — not stored; maps source_type to human-readable label
    source_category_label: str = ""
    # Injected by router — not a DB column
    active_followup: Optional[FollowupOut] = None

    model_config = {"from_attributes": True}

    _validate_appointment_time_slot = field_validator("appointment_time_slot", mode="before")(
        _normalize_time_slot
    )
    _validate_duration_minutes = field_validator("estimated_job_duration_minutes", mode="before")(
        _normalize_duration_minutes
    )

    @field_validator("field_sources", mode="before")
    @classmethod
    def _parse_field_sources(cls, v: Any) -> Any:
        """field_sources is stored as a JSON string in the DB; deserialize it here."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (ValueError, TypeError):
                return None
        return v

    @field_validator("move_date_options", mode="before")
    @classmethod
    def _parse_move_date_options(cls, v: Any) -> Any:
        """move_date_options is stored as a JSON array string in the DB; deserialize it here."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (ValueError, TypeError):
                return None
        return v

    @field_validator("quote_modifiers", mode="before")
    @classmethod
    def _parse_quote_modifiers(cls, v: Any) -> Any:
        """quote_modifiers is stored as a JSON array string in the DB; deserialize it here."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (ValueError, TypeError):
                return None
        return v

    @model_validator(mode="after")
    def _set_source_label(self) -> "LeadOut":
        key = self.source_type.value if hasattr(self.source_type, "value") else str(self.source_type)
        self.source_category_label = _SOURCE_LABELS.get(key, key)
        return self


class LeadDetailOut(LeadOut):
    events: list[LeadEventOut] = []
    screenshots: list[ScreenshotOut] = []

File: app/frontend/src/types/lead.ts
export type LeadStatus =
  | 'new'
  | 'in_review'
  | 'replied'
  | 'waiting_on_customer'
  | 'ready_for_quote'
  | 'ready_for_booking'
  | 'escalated'
  | 'booked'
  | 'released'
  | 'lost'

export type LeadSourceType =
  | 'thumbtack_api'
  | 'thumbtack_screenshot'
  | 'yelp_screenshot'
  | 'google_screenshot'
  | 'website_form'
  | 'manual'

export type ServiceType = 'moving' | 'hauling' | 'both' | 'unknown'

export interface LeadEvent {
  id: string
  lead_id: string
  event_type: string
  from_status: string | null
  to_status: string | null
  actor: string | null
  note: string | null
  created_at: string
}

export interface Screenshot {
  id: string
  lead_id: string
  original_filename: string
  stored_path: string  // prefix with /uploads/ to build URL
  file_size: number
  ocr_status: 'pending' | 'done' | 'failed' | null
  screenshot_type: string  // "intake" | "correspondence" | "before_job" | "after_job"
  created_at: string
}

export interface OcrField {
  field: string
  value: string
  confidence: 'high' | 'medium' | 'low'
}

export interface OcrResult {
  id: string
  screenshot_id: string
  raw_text: string | null
  /** JSON string — parse with JSON.parse() to get OcrField[] */
  extracted_fields: string | null
  model_used: string | null
  created_at: string
}

export interface QuoteModifier {
  amount: number
  note: string
}

export interface AiReviewSections {
  // Action-first
  a_next_message: string
  b_call_plan: string
  c_behavior_class: string
  d_transport_path: string
  e_escalation_note: string
  // Pricing & Control (internal only — never share with customer)
  f_pricing_band: string
  g_band_position: string
  h_friction_points: string
  i_sayability_check: string
  j_quote_style: string
  k_quote_source_label: string
  l_pricing_guidance: string
  // Support & Context
  m_quick_read: string
  n_pattern_anchor: string
  o_branch_replies: string
}

export interface AiReview {
  id: string
  lead_id: string
  model_used: string
  prompt_version: string
  grounding_source: string | null
  sections: AiReviewSections
  input_snapshot: Record<string, unknown>
  created_at: string
  actor: string | null
}

export interface IngestResult {
  lead: Lead
  extraction: OcrResult | null
  auto_applied_fields: string[]
}

export interface WebhookIngestResult {
  lead: Lead | null
  created: boolean
  was_duplicate: boolean
  message: string | null
}

export interface Lead {
  id: string
  city_id: string
  city_name: string | null
  city_slug: string | null
  source_type: LeadSourceType
  source_reference_id: string | null
  raw_payload: string | null
  status: LeadStatus
  urgency_flag: boolean
  // null for ingest stubs not yet confirmed by OCR or the facilitator
  customer_name: string | null
  customer_phone: string | null
  service_type: ServiceType
  job_location: string | null
  job_origin: string | null
  job_destination: string | null
  job_date_requested: string | null
  job_date_end: string | null
  appointment_time_slot: string | null
  estimated_job_duration_minutes: number | null
  scope_notes: string | null
  // Provenance dict: field_name → "ocr" | "edited" (absence = manually entered)
  field_sources: Record<string, 'ocr' | 'edited'> | null
  notes: string | null
  ingested_by: string | null
  assigned_to: string | null
  created_at: string
  acknowledged_at: string | null
  updated_at: string
  dispatched_at: string | null
  started_at: string | null
  // Slice 8
  move_distance_miles: number | null
  load_stairs: number | null
  unload_stairs: number | null
  move_size_label: string | null
  move_type: string | null
  move_date_options: string[] | null
  accept_and_pay: boolean
  contact_status: 'locked' | 'unlocked'
  acknowledgment_sent: boolean
  quote_context: string | null
  quoted_price_total: number | null
  quote_modifiers: QuoteModifier[] | null
  job_address: string | null
  source_category_label: string
  events?: LeadEvent[]
  screenshots?: Screenshot[]
}

export interface LeadCreate {
  city_id?: string
  source_type: LeadSourceType
  source_reference_id?: string
  raw_payload?: string
  customer_name: string
  customer_phone?: string
  service_type: ServiceType
  job_location?: string
  job_date_requested?: string
  appointment_time_slot?: string
  estimated_job_duration_minutes?: number
  notes?: string
  assigned_to?: string
  urgency_flag?: boolean
}

export interface LeadUpdate {
  customer_name?: string
  customer_phone?: string | null
  service_type?: ServiceType
  job_location?: string | null
  job_origin?: string | null
  job_destination?: string | null
  job_date_requested?: string | null
  job_date_end?: string | null
  appointment_time_slot?: string | null
  estimated_job_duration_minutes?: number | null
  scope_notes?: string | null
  urgency_flag?: boolean
  assigned_to?: string | null
  notes?: string | null
  // Slice 8
  move_distance_miles?: number | null
  load_stairs?: number | null
  unload_stairs?: number | null
  move_size_label?: string | null
  move_type?: string | null
  move_date_options?: string[] | null
  acknowledgment_sent?: boolean
  quote_context?: string | null
  job_address?: string | null
}

export interface ChatMessage {
  id: string
  lead_id: string
  ai_review_id: string | null
  role: 'user' | 'assistant'
  content: string
  created_at: string
}

export interface ChatResponse {
  messages: ChatMessage[]
  quote_context_update?: string | null
}

export interface Settings {
  t1_minutes: number
  t2_minutes: number
  quiet_hours_start: string
  quiet_hours_end: string
  quiet_hours_enabled: boolean
  primary_sms: string
  primary_email: string
  backup_name: string
  backup_sms: string
  backup_email: string
}

export interface SettingsPatch {
  t1_minutes?: number
  t2_minutes?: number
  quiet_hours_start?: string
  quiet_hours_end?: string
  quiet_hours_enabled?: boolean
  primary_sms?: string
  primary_email?: string
  backup_name?: string
  backup_sms?: string
  backup_email?: string
}

export interface TestAlertRequest {
  channel: 'sms' | 'email'
  recipient: 'primary' | 'backup'
}

export interface TestAlertResult {
  sent: boolean
  reason?: string | null
}
